### PR TITLE
HTTPProxy: Add validation check

### PR DESCRIPTION
**Pull request by abhide**
Add validation to ensure that Passthrough and Fallback aren't specified
at the same time.

Fixes: #3197
Signed-off-by: Amey Bhide <amey15@gmail.com>

### CODE REVIEW

File: internal/dag/httpproxy_processor.go [hunk 1]
```diff
@@ -144,6 +144,10 @@ func (p *HTTPProxyProcessor) computeHTTPProxy(proxy *contour_api_v1.HTTPProxy) {
 
 	var tlsEnabled bool
 	if tls := proxy.Spec.VirtualHost.TLS; tls != nil {
+		/*if tls.Passthrough && tls.EnableFallbackCertificate {
```
Comment (abhide) [inline internal/dag/httpproxy_processor.go:147]: @jpeach Where should I add a unit test for this? 

```
./internal/xdscache/v3/route_test.go
./internal/envoy/v3/route_test.go
./internal/featuretests/v3/route_test.go
```

Comment (youngnick) [reply internal/dag/httpproxy_processor.go:147]: `./internal/dag/builder_test.go` is actually the spot to go, you'll see tests for all the processors in there. Probably we need to fix that up, but that's the spot to start.

File: internal/dag/httpproxy_processor.go [hunk 2]
```diff
@@ -144,6 +144,10 @@ func (p *HTTPProxyProcessor) computeHTTPProxy(proxy *contour_api_v1.HTTPProxy) {
...
+			validCond.AddError(contour_api_v1.ConditionTypeAuthError, "TLSIncompatibleFeatures",
```
Comment (sunjayBhatia) [inline internal/dag/httpproxy_processor.go:148]: nit: this should probably end up being `contour_api_v1.ConditionTypeTLSError`

File: internal/dag/httpproxy_processor.go (hunk 1 above)
Comment (sunjayBhatia) [reply internal/dag/httpproxy_processor.go:147]: Do we have somewhere we test for error conditions we set while processing objects? I haven't been able to find much

Comment (skriss) [reply internal/dag/httpproxy_processor.go:147]: > Do we have somewhere we test for error conditions we set while processing objects? I haven't been able to find much

`internal/dag/status_test.go` has some of those, and then `internal/featuretests` do as well.

Comment (abhide) [reply internal/dag/httpproxy_processor.go]: Thanks. Added a test to `status_test`

File: internal/dag/httpproxy_processor.go (hunk 2 above)
Comment (abhide) [reply internal/dag/httpproxy_processor.go]: Thanks. Fixed

File: internal/dag/status_test.go [hunk 3]
```diff
@@ -2369,4 +2369,75 @@ func TestDAGStatus(t *testing.T) {
 		},
 	})
 
+	// issue 3197: Fallback and passthrough HTTPProxy directive should emit a config error
+	invalidFallbackCertificateSvc := &contour_api_v1.HTTPProxy{
```
Comment (skriss) [inline internal/dag/status_test.go:2373]: Maybe `tlsPassthroughAndFallback` would be a more descriptive name for the variable

File: internal/dag/status_test.go [hunk 4]
```diff
@@ -2369,4 +2369,75 @@ func TestDAGStatus(t *testing.T) {
...
+				}},
+				Services: []contour_api_v1.Service{{
+					Name: "home",
+					Port: 8080,
+				}},
+			}},
+		},
+	}
+
+	run(t, "invalid TLS", testcase{
```
Comment (skriss) [inline internal/dag/status_test.go:2399]:
```suggestion
	run(t, "TLS with passthrough and fallback cert enabled is invalid", testcase{
```

File: internal/dag/status_test.go [hunk 5]
```diff
@@ -2369,4 +2369,75 @@ func TestDAGStatus(t *testing.T) {
...
+		objs: []interface{}{invalidFallbackCertificateSvc, fixture.ServiceRootsHome},
+		want: map[types.NamespacedName]contour_api_v1.DetailedCondition{
+			{Name: invalidFallbackCertificateSvc.Name, Namespace: invalidFallbackCertificateSvc.Namespace}: fixture.NewValidCondition().
+				WithGeneration(invalidFallbackCertificateSvc.Generation).WithError(
+				contour_api_v1.ConditionTypeTLSError, "TLSIncompatibleFeatures",
+				`Spec.VirtualHost.TLS: both Passthrough and enableFallbackCertificate were specified`,
+			),
+		},
+	})
+	validFallbackCertificateSvc := &contour_api_v1.HTTPProxy{
```
Comment (skriss) [inline internal/dag/status_test.go:2409]:
```suggestion
	tlsPassthrough := &contour_api_v1.HTTPProxy{
```

File: internal/dag/status_test.go [hunk 6]
```diff
@@ -2369,4 +2369,75 @@ func TestDAGStatus(t *testing.T) {
...
+				}},
+				Services: []contour_api_v1.Service{{
+					Name: "home",
+					Port: 8080,
+				}},
+			}},
+		},
+	}
+
+	run(t, "valid TLS", testcase{
```
Comment (skriss) [inline internal/dag/status_test.go:2435]:
```suggestion
	run(t, "valid TLS passthrough", testcase{
```

File: internal/dag/status_test.go [hunk 7]
```diff
@@ -2369,4 +2369,75 @@ func TestDAGStatus(t *testing.T) {
...
+		objs: []interface{}{validFallbackCertificateSvc, fixture.ServiceRootsHome},
```
Comment (skriss) [inline internal/dag/status_test.go:2436]:
```suggestion
		objs: []interface{}{tlsPassthrough, fixture.ServiceRootsHome},
```

File: internal/dag/status_test.go [hunk 8]
```diff
@@ -2369,4 +2369,75 @@ func TestDAGStatus(t *testing.T) {
...
+		want: map[types.NamespacedName]contour_api_v1.DetailedCondition{
+			{Name: invalidFallbackCertificateSvc.Name, Namespace: invalidFallbackCertificateSvc.Namespace}: fixture.NewValidCondition().
```
Comment (skriss) [inline internal/dag/status_test.go:2438]:
```suggestion
			{Name: tlsPassthrough.Name, Namespace: tlsPassthrough.Namespace}: fixture.NewValidCondition().
```

File: internal/dag/status_test.go [hunk 9]
```diff
@@ -2369,4 +2369,75 @@ func TestDAGStatus(t *testing.T) {
...
+				WithGeneration(invalidFallbackCertificateSvc.Generation).
```
Comment (skriss) [inline internal/dag/status_test.go:2439]:
```suggestion
				WithGeneration(tlsPassthrough.Generation).
```

**Review by skriss — CHANGES_REQUESTED** [review]
Thanks for the PR @abhide! Had a few minor comments.

File: internal/dag/status_test.go (hunk 3 above)
Comment (abhide) [reply internal/dag/status_test.go:2373]: Done

File: internal/dag/status_test.go (hunk 4 above)
Comment (abhide) [reply internal/dag/status_test.go:2399]: Done

File: internal/dag/status_test.go (hunk 5 above)
Comment (abhide) [reply internal/dag/status_test.go:2409]: Done

File: internal/dag/status_test.go (hunk 6 above)
Comment (abhide) [reply internal/dag/status_test.go:2435]: Done

File: internal/dag/status_test.go (hunk 7 above)
Comment (abhide) [reply internal/dag/status_test.go:2436]: Done

File: internal/dag/status_test.go (hunk 8 above)
Comment (abhide) [reply internal/dag/status_test.go:2438]: Done

File: internal/dag/status_test.go (hunk 9 above)
Comment (abhide) [reply internal/dag/status_test.go:2439]: Done

**Comment by abhide** [comment]
@skriss addressed your comments

**Review by skriss — APPROVED** [review]
LGTM, thanks @abhide!

**Review by youngnick — APPROVED** [review]

**Pull request merged**